modules/core/src/main/java/org/apache/ignite/internal/processors/cache/QueryCursorImpl.java

```diff
@@ -106,7 +129,8 @@ public QueryCursorImpl(Iterable<T> iterExec) {
                     throw new IgniteException(e);
                 }
             }
-        }
+        } else if (cancellable && cancel != null)
```

Comment:
This code looks really complex, there are lots of possible races between iterator initialization and concurrent cancellation. I will not be surprised if we have bugs in `iter.close()` and `cancel.apply()` implementations.

Moreover `iter` field is not even volatile, so can find partially constructed object here.

Also cancel can be called multiple times here.

Why do we have this `cancellable` flag? It seems that it makes no difference and everything depends on `iter` being initialized.

Lets discuss this.